### DISCUSSION TITLE
请问umi4还支持多config目录下多环境配置吗？

### DISCUSSION THREAD

**Discussion by zhouzyc** [umijs/umi#8341] [Q&A]
刚才试了下，从umi3升级到umi4.配置如下图，.local.ts在任何环境都会被执行，导致配置被覆盖无效， 在umi4文档中并没有相关说明，请问还支持这样写吗。
<img width="1232" alt="image" src="https://user-images.githubusercontent.com/18420689/176838114-3f3a190c-63ae-4d19-bacb-f13e7a22cb13.png">

**Comment by fz6m** [comment]
合并顺序：

 - `config.ts`

 - `config.${UMI_ENV}.ts`

 - `config.${dev | prod}.ts`

 - `config.${dev | prod}.${UMI_ENV}.ts`

 - `config.local.ts`

越向下的越具体，优先级更高，高优的可以往下移动。

**Reply by zhouzyc** [reply]
> 合并顺序：
> 
> * `config.ts`
> * `config.${UMI_ENV}.ts`
> * `config.${dev | prod}.ts`
> * `config.${dev | prod}.${UMI_ENV}.ts`
> * `config.local.ts`
> 
> 越向下的越具体，优先级更高，高优的可以往下移动。

非常感谢

**Reply by HaishengLiang** [reply]
反对 config.local.ts 被作为 dev 以外的配置项

**Reply by woshiji1987** [reply]
max setup对config目录下的config.ts 运行时引入插件无效，不会自动引入插件代码，配置放到外层umirc.js就没问题，这是什么情况？

**Reply by ghost** [reply]
> 反对 config.local.ts 被作为 dev 以外的配置项

有的项目本地打包直接推送到server 这样的话 就不行了,哎 不知道为啥这样design

**Reply by BH-M87** [reply]
这个合并顺序和umi@3有所不同了。

另外，明确指定了环境变量的情况下，优先级应该高于dev或者prod，所以合理的顺序应该是：

- config.ts
- config.${dev | prod}.ts
- config.${UMI_ENV}.ts
- config.${dev | prod}.${UMI_ENV}.ts
- config.local.ts

**Reply by peter064226** [reply]
https://github.com/umijs/umi/discussions/10448  请问config.test.ts这个配置如何使其生效呢？

**Reply by visla-perlin** [reply]
这设计真的好迷

**Reply by visla-perlin** [reply]
如果我没有自动化部署的环境，只要config.local.ts存在，我永远无法在本地执行prod环境的发布，因为最终都会被config.local.ts覆盖。如果我要发布到prod，那么config.local.ts也是绝对不能上传到的，这又加了一道红线。

**Reply by fz6m** [reply]
默认情况下本地配置 `config.local.ts` 永远优先，同时需要搭配 `.gitignore` 忽略这个文件，从而实现只有本地生效的配置效果，90% 的情况下，我们不会本地构建生产环境的包，如果有此类需求，可以不用 local ，使用 `config.dev.ts` 做本地开发配置。

**Reply by visla-perlin** [reply]
我有三个环境，本地环境 使用config.local.ts  ,prod 环境使用config.ts，可唯独dev测试环境的 config.dev.ts没法用，因为会被prod覆盖了

**Reply by Deja-vuuu** [reply]
这为啥不借鉴一下nextjs的环境变量配置，通俗易懂。翻了很多都说这玩意设计的难以理解。😂

**Reply by iversonwool** [reply]
大佬，能发一下这个配置合并的源码链接吗，我想看看，感谢。

**Reply by zuozuoking** [reply]
我无法配置打包环境 我 配置的打包地址不生效  他永远就认那个 .env 里面的地址
![image](https://github.com/user-attachments/assets/9f9a09c1-beda-47ff-88f6-0d8bee63a0a7)
![image](https://github.com/user-attachments/assets/a423ddab-2742-4c86-af78-b1a30801bad2)

**Reply by zuozuoking** [reply]
可以给我一个解决办法吗 谢谢

**Comment by ZeroJsus** [comment]
这个从下往上的优先级 让umi3 直接转过来的有点难受

**Reply by BH-M87** [reply]
基本上，无法直接升级。找到这些坑并修改投入了很多时间😭

**Reply by ZeroJsus** [reply]
> 

是的 我还有chainwebpack的问题 之前在umi3写的插件 可以使用，到了4使用都直接配置文件错误了

**Reply by BH-M87** [reply]
我在业务的间隙中升级的，前前后后2周时间才解决完所有问题。回头看看都不好和老板交代了，投入的时间和产出，roi不太好看。

**Reply by BH-M87** [reply]
还有个history的问题，render前不要调用history.push，不然会报错。配置文件的优先级问题也要注意下。css modules的配置似乎也不一样了哦。

**Reply by SilentFlute** [reply]
> 还有个history的问题，render前不要调用history.push，不然会报错。配置文件的优先级问题也要注意下。css modules的配置似乎也不一样了哦。

'render前不要调用history.push，不然会报错', 是这样了, 我也遇到了, 在做自定义渲染不走全局`layout`的时候遇到的, 报错信息如下:
```
Warning: Cannot update a component (`BrowserRoutes`) while rendering a different component (`LayoutWrapper`).
To locate the bad setState() call inside `LayoutWrapper`, follow the stack trace as described in
https://reactjs.org/link/setstate-in-render
```

**Reply by BH-M87** [reply]
> > 还有个history的问题，render前不要调用history.push，不然会报错。配置文件的优先级问题也要注意下。css modules的配置似乎也不一样了哦。
> 
> 'render前不要调用history.push，不然会报错', 是这样了, 我也遇到了, 在做自定义渲染不走全局`layout`的时候遇到的, 报错信息如下:
> 
> ```
> Warning: Cannot update a component (`BrowserRoutes`) while rendering a different component (`LayoutWrapper`).
> To locate the bad setState() call inside `LayoutWrapper`, follow the stack trace as described in
> https://reactjs.org/link/setstate-in-render
> ```

是的。这个在别的框架里面一般是不会遇到的。这里是因为他history的初始化的时机做的不太好。它实在render后去create的。其实这个一开始就能初始化掉。

**Reply by fz6m** [reply]
`history` 是需要依赖动态参数来进行创建的，在创建前，这些参数可能在运行时被修改，如 qiankun 微前端可能会修订路由类型，所以 render 后才能使用 history 。

在 umi 中是有独立 history 可以在任意位置使用的，而在 react router 中，只能在 hooks 中使用，从时机上他们都在渲染后才能使用。

**Reply by BH-M87** [reply]
> `history` 是需要依赖动态参数来进行创建的，在创建前，这些参数可能在运行时被修改，如 qiankun 微前端可能会修订路由类型，所以 render 后才能使用 history 。
> 
> 在 umi 中是有独立 history 可以在任意位置使用的，而在 react router 中，只能在 hooks 中使用，从时机上他们都在渲染后才能使用。

umi3里面是会默认初始化一个，然后再替换的。就不会有问题。umi4里面逻辑改了，一开始没初始化，就不行。可以自己写代码绕过去就是。

**Comment by lecd** [comment]
最终解决了吗

**Reply by sumbyKB** [reply]
默认的最好不要去命名修改它，比如不要去定义以及修改 config.dev.ts 或者 config.prod.ts，建议直接命名一个 config.online.ts（自己想要的环境命名）， 跑项目的时候     "start:online": "cross-env UMI_ENV=online umi dev" 就会执行指定的 online 配置了

**Reply by SilentFlute** [reply]
为什么我用你这个方法无法修改UMI_ENV呢:
`/config/config.dev.ts`:
```
  define: {
    baseUrl: 'dev baseUrl'
  }
```
`/config/config.test.ts`:
```
  define: {
    baseUrl: 'test baseUrl'
  }
```
`/config/config.online.ts`:
```
  define: {
    baseUrl: 'online baseUrl'
  }
```
```
    "dev": "cross-env UMI_ENV=dev && umi dev",
    "test": "cross-env UMI_ENV=test && umi dev",
    "online": "cross-env UMI_ENV=online && umi dev",
```
上面那3个命令都走的是`/config/config.dev.ts`里面的配置, 我的`baseUrl`始终是`dev baseUrl`, 这是为何啊?
然后我也没找到能修改`UMI_ENV`的方法

**Reply by SilentFlute** [reply]
看了文档: [UMI_ENV](https://umijs.org/docs/guides/env-variables#umi_env) 发现貌似没法修改`UMI_ENV`的值, 那是如何能够加载`config.${UMI_ENV}.ts`的呢, 怎么区分当前是`dev`还是`test`还是`production`的呢?

**Reply by BH-M87** [reply]
可能是个bug？删除掉config.dev.ts再试试？

**Reply by fz6m** [reply]
仔细看上文描述的优先级顺序，`config.dev.ts` 优先级更高。

**Reply by BH-M87** [reply]
> 仔细看上文描述的优先级顺序，`config.dev.ts` 优先级更高。

所以这里，不要加`config.dev.ts`就没事。不然如果有了这个文件，`UMI_ENV`设置啥都没用。我觉得这个设计并不太合理。去年8月份就提过了。
<img width="945" alt="CleanShot 2023-01-29 at 16 28 54@2x" src="https://user-images.githubusercontent.com/16903266/215314648-eaac8bd4-db47-4020-933f-f0529976d687.png">

**Reply by SilentFlute** [reply]
我试了, 还是不行, 只有一个和`UMI_ENV`值匹配的配置文件, 此时页面上直接报错了:
`ReferenceError: baseUrl is not defined`
这个是最小可复现的仓库: https://github.com/SilentFlute/umi4-simple_multiple-env-sample
看起来是配置丢失了, 也就是我的这个配置文件并没有生效...

**Reply by BH-M87** [reply]
> 我试了, 还是不行, 只有一个和`UMI_ENV`值匹配的配置文件, 此时页面上直接报错了: `ReferenceError: baseUrl is not defined` 这个是最小可复现的仓库: https://github.com/SilentFlute/umi4-simple_multiple-env-sample 看起来是配置丢失了, 也就是我的这个配置文件并没有生效...

我看了下，是真的没生效。
<img width="1029" alt="CleanShot 2023-02-01 at 10 41 25@2x" src="https://user-images.githubusercontent.com/16903266/215932264-a0ad4369-e67d-49d6-93ce-4e38556e4e84.png">
这个配置只有在config.ts里面才能生效。

**Reply by SilentFlute** [reply]
那看来似乎是bug了, 大佬抽空看一眼? @fz6m

**Reply by fz6m** [reply]
只有存在 `.umirc.ts` or `config.ts` 的时候，特定环境配置文件才会生效，不可以只特定环境。

可以加一个空的 `config.ts` ：

```ts
export default {}
```

**Reply by BH-M87** [reply]
> 只有存在 `.umirc.ts` or `config.ts` 的时候，特定环境配置文件才会生效，不可以只特定环境。
> 
> 可以加一个空的 `config.ts` ：
> 
> ```ts
> export default {}
> ```

这里合适的方案应该是如果没有`config.ts`，框架也能处理的吧。这可能还是一个bug？

**Reply by fz6m** [reply]
设计就是这样的，没有主配置文件，附属文件不会生效。

**Reply by gitdust** [reply]
> 只有存在 `.umirc.ts` or `config.ts` 的时候，特定环境配置文件才会生效，不可以只特定环境。
> 
> 可以加一个空的 `config.ts` ：
> 
> ```ts
> export default {}
> ```

不是 .umirc config 二选一吗？

**Reply by Deja-vuuu** [reply]
> 设计就是这样的，没有主配置文件，附属文件不会生效。

这设计我感觉很离谱

**Reply by Aqurius-Hunter** [reply]
这个设计思路实在是看不懂，建议用简单一目了然的方式。UMI_ENV 等于什么就用什么配置不是最好吗~

**Comment by croatialu** [comment]
给人一种....


算了， 我目前的多环境配置

``` json
{
  "script": {
    "build": "cross-env UMI_ENV=prod max build",
    "build:test": "cross-env UMI_ENV=test max build",
  }
}

```

然后在 `/config/config.prod.prod.ts` 和 `/config/config.prod.test.ts` 中写自己的一些配置

不要掉了中间的一个 `prod`...

**Reply by visla-perlin** [reply]
想要自己的配置文件，还得先绕过默认的配置文件，😄

**Reply by zuozuoking** [reply]
这样写 我的还是不生效  
![image](https://github.com/user-attachments/assets/aef6cb80-1f04-43f6-a929-229c15cbaaae)
![image](https://github.com/user-attachments/assets/9d3bcefc-2aeb-48c9-aabf-b13e0e37fb45)

**Reply by SilentFlute** [reply]
我的这么写是生效的, 这个项目是之前一个同事搭的, 用的@umijs/max, 我只是参与迭代, 已经上线, 目前没什么问题:
<img width="1273" alt="image" src="https://github.com/user-attachments/assets/1849d795-a5be-4e84-89b9-34ed92b6a505">

**Comment by fz6m** [comment]
如果不想配置文件过多复杂化，可以在一个配置文件 `config.ts` 里判断 `process.env.UMI_ENV` 和 `process.env.NODE_ENV` 的值，从而更改不同的配置即可。

**Reply by KaiZheng2020** [reply]
config.ts 识别不了 process.env.UMI_ENV 和 process.env.NODE_ENV

**Reply by SilentFlute** [reply]
可以啊, 我刚试了可以拿到它们的值的啊
